File: src/piwardrive/map/vector_tile_customizer.py
"""Utilities for customizing MBTiles vector tile sets."""

from __future__ import annotations

import json
import logging
import os
import sqlite3


def apply_style(
    path: str,
    *,
    style_path: str | None = None,
    name: str | None = None,
    description: str | None = None,
) -> None:
    """Update ``path`` metadata with the given style and fields."""
    if not os.path.exists(path):
        raise FileNotFoundError(path)

    style_json: str | None = None
    if style_path:
        try:
            with open(style_path, "r", encoding="utf-8") as fh:
                style_json = fh.read()
                json.loads(style_json)  # validate
        except OSError as exc:
            raise FileNotFoundError(style_path) from exc
        except json.JSONDecodeError as exc:
            raise ValueError(f"Invalid style JSON in {style_path}: {exc}") from exc

    with sqlite3.connect(path) as db:
        db.execute("CREATE TABLE IF NOT EXISTS metadata (name TEXT, value TEXT)")
        for key, value in (
            ("name", name),
            ("description", description),
            ("style", style_json),
        ):
            if value is not None:
                db.execute(
                    "INSERT OR REPLACE INTO metadata (name, value) VALUES (?, ?)",
                    (key, value),
                )
        db.commit()


def build_mbtiles(folder: str, output: str) -> None:
    """Create an MBTiles file from ``folder`` of XYZ PBF tiles."""
    if not os.path.isdir(folder):
        raise FileNotFoundError(folder)

    with sqlite3.connect(output) as db:
        db.executescript(
            """
            CREATE TABLE IF NOT EXISTS tiles (
                zoom_level INTEGER,
                tile_column INTEGER,
                tile_row INTEGER,
                tile_data BLOB,
                UNIQUE (zoom_level, tile_column, tile_row)
            );
            CREATE TABLE IF NOT EXISTS metadata (name TEXT, value TEXT);
            """
        )
        for root, _, files in os.walk(folder):
            for f in files:
                if not f.endswith(".pb"):
                    continue
                zxy = os.path.relpath(os.path.join(root, f), folder).split(os.sep)
                if len(zxy) != 3:
                    continue
                z, x, yext = zxy
                y, _ = os.path.splitext(yext)
                try:
                    z_i = int(z)
                    x_i = int(x)
                    y_i = int(y)
                except ValueError:
                    continue
                tile_row = (2**z_i - 1) - y_i  # TMS
                try:
                    with open(os.path.join(root, f), "rb") as fh:
                        data = fh.read()
                except OSError as exc:
                    logging.exception("Failed to read tile %s: %s", f, exc)
                    continue
                db.execute(
                    (
                        "INSERT OR REPLACE INTO tiles ",
                        "(zoom_level, tile_column, tile_row, tile_data) ",
                        "VALUES (?, ?, ?, ?)",
                    ),
                    (z_i, x_i, tile_row, data),
                )
        db.commit()

File: webui/src/vectorTileCustomizer.js
import fs from 'fs';
import path from 'path';
import os from 'os';
import { execFileSync } from 'child_process';

export function applyStyle(dbPath, { stylePath, name, description } = {}) {
  const lines = [
    'CREATE TABLE IF NOT EXISTS metadata (name TEXT, value TEXT);',
  ];
  if (name != null)
    lines.push(
      `INSERT OR REPLACE INTO metadata (name,value) VALUES ('name','${name}')`
    );
  if (description != null)
    lines.push(
      `INSERT OR REPLACE INTO metadata (name,value) VALUES ('description','${description}')`
    );
  if (stylePath) {
    const style = fs.readFileSync(stylePath, 'utf-8');
    JSON.parse(style);
    lines.push(
      `INSERT OR REPLACE INTO metadata (name,value) VALUES ('style','${style.replace(/'/g, "''")}')`
    );
  }
  execFileSync('sqlite3', [dbPath], { input: lines.join('\n') });
}

export function buildMbtiles(folder, output) {
  if (!fs.existsSync(folder) || !fs.statSync(folder).isDirectory()) {
    throw new Error('Folder not found');
  }
  const sql = [];
  sql.push(
    'CREATE TABLE IF NOT EXISTS tiles (zoom_level INTEGER, tile_column INTEGER, tile_row INTEGER, tile_data BLOB, UNIQUE (zoom_level,tile_column,tile_row));'
  );
  sql.push('CREATE TABLE IF NOT EXISTS metadata (name TEXT, value TEXT);');
  for (const rootZ of fs.readdirSync(folder)) {
    const zDir = path.join(folder, rootZ);
    if (!fs.statSync(zDir).isDirectory()) continue;
    for (const xEntry of fs.readdirSync(zDir)) {
      const xDir = path.join(zDir, xEntry);
      if (!fs.statSync(xDir).isDirectory()) continue;
      for (const file of fs.readdirSync(xDir)) {
        if (!file.endsWith('.pbf')) continue;
        const yName = path.basename(file, '.pbf');
        const z = parseInt(rootZ, 10);
        const x = parseInt(xEntry, 10);
        const y = parseInt(yName, 10);
        if (Number.isNaN(z) || Number.isNaN(x) || Number.isNaN(y)) continue;
        const tileRow = Math.pow(2, z) - 1 - y;
        const data = fs.readFileSync(path.join(xDir, file));
        const hex = data.toString('hex');
        sql.push(
          `INSERT OR REPLACE INTO tiles (zoom_level,tile_column,tile_row,tile_data) VALUES (${z},${x},${tileRow},X'${hex}')`
        );
      }
    }
  }
  const tmp = path.join(os.tmpdir(), `mb-${Date.now()}.sql`);
  fs.writeFileSync(tmp, sql.join('\n'));
  execFileSync('sqlite3', [output], { input: fs.readFileSync(tmp) });
  fs.unlinkSync(tmp);
}
